client/packages/plugins/README.md

```diff
@@ -1,12 +1,310 @@
+
+See the readme on plugin validation for details of how to sign plugins and the process of loading and validating plugins.
+
+### Plugin definitions
+
+Each plugin is defined in a `plugin.json` file which has the following structure
+
+```
+{
+  "name": "StockDonor",
+  "version": "1.0.0",
+  "components": [
+    {
+      "type": "StockEditForm",
+      "localModule": "StockDonorEditInput",
+      "module": "StockDonorEdit"
+    }
+  ],
+  "columns": [
+    {
+      "type": "Stock",
+      "module": "StockDonorColumn"
+    }
+  ],
+  "dependencies": {
+    "omSupplyVersion": "8.0.0"
+  }
+}
+```
+
+The following types of plugins are currently supported:
+
+**ColumnPlugin**
+- `Stock`
+
+**ComponentPlugin**
+- `StockEditForm`
+- `Stock`
+- `InboundShipmentAppBar`
+- `Dashboard`
+
+
+**Plugin events**
+- `onSaveStockEditForm`
+- `onChangeStockEditForm`
+
+
+The components array lists all of the components exported by the plugin, the columns array lists all of the columns. The `name` property must match the name of the directory, as this is used to link the exported js file. See the examples for how this is implemented in the `webpack.config.js` file.
+
+### Types of plugin
+
+There are two types of plugin:
+**ColumnPlugin**
+
+The column plugin exports a `ColumnDefinition` for the given data type.
+This is defined in the plugin provider by the following object shape:
+
+```
+  type: ColumnPluginType;
+  column: () => Promise<ColumnDefinition<T>>;
+  module: string;
+  pluginName: string;
+```
+
+**ComponentPlugin**
+A component plugin exports a react component for the given data type. 
+This is defined in the plugin provider by the following object shape:
+
+
+```
+  type: ComponentPluginType;
+  component: () => Promise<PluginModule<T>>;
+  module: string;
+  localModule?: string;
+  pluginName: string;
+
+```
+
+There is a difference in behaviour when the app is running in development or production mode. See the **Plugin structure** section for more on this.
+A `PluginProvider` stores a list of available plugins. A component within the app can retrieve a list of applicable plugins by calling the `getComponentPlugins` or `getColumnPlugins` method, passing in the type of plugin required. Even simpler, use the hook provided to fetch components like this:
+
+```
+  const pluginButtons = usePluginElements({
+    type: 'InboundShipmentAppBar',
+    data,
+  });
+```
+
+which are then rendered directly:
+
+```
+{pluginButtons}
+```
+
+Similarly, the columns have a hook provided as well:
+
+```
+  const pluginColumns = usePluginColumns<StockLineRowFragment>({
+    type: 'Stock',
+  });
+```
+
+```
+  const columnDefinitions: ColumnDescription<StockLineRowFragment>[] = [
+    [standard column definitions go here],
+    ...pluginColumns,
+  ];
+
+...
+
+  const columns = useColumns<StockLineRowFragment>(
+    columnDefinitions,
+    {
+      sortBy,
+      onChangeSortBy: updateSortQuery,
+    },
+    [sortBy, pluginColumns]
+  );
+```
+
+
+### Events
```

Comment:
I got really stuck on this section, I looked at examples, which helped (btw maybe good to have a link to the example branch at the top ?).

I added to the code to try to clarify the confusion: https://github.com/openmsupply/open-msupply/compare/2372-plugins-readme...2372-plugins-readme(Add-to-event-example-code)

You'll notice that I just added Event(params), I wasn't sure what to put in the calling method, what should the params be ? Why does it have stockline.id in the examples and what does that do ?

Reply:
Btw apart from that, I think i understood the whole document in one go thanks !